runtime/src/tokio.rs

```diff
@@ -501,23 +504,51 @@ pub struct Blob {
     partition: String,
     name: Vec<u8>,
 
-    file: fs::File,
+    // Files must be seeked prior to any read or write operation and are thus
+    // not safe to concurrently interact with. If we switched to mmaping files
+    // we could remove this lock.
+    //
+    // We also track the virtual file size because metadata isn't updated until
+    // the file is synced (not to mention it is a lot less fs calls).
+    file: Arc<AsyncMutex<(fs::File, u64)>>,
 }
 
 impl Blob {
-    fn new(metrics: Arc<Metrics>, partition: String, name: &[u8], file: fs::File) -> Self {
+    fn new(
+        metrics: Arc<Metrics>,
+        partition: String,
+        name: &[u8],
+        file: fs::File,
+        len: u64,
+    ) -> Self {
         metrics.open_blobs.inc();
         Self {
             metrics,
             partition,
             name: name.into(),
-            file,
+            file: Arc::new(AsyncMutex::new((file, len))),
+        }
+    }
```

Comment:
This is required because tokio requires a file to be seeked prior to access. When this is replaced with `mmap'ing`, we can make this much more efficient: #154